### PR TITLE
feat(reconciler): Shared reconciliation loop logic for controllers

### PR DESCRIPTION
**Pull request by abhijith-darshan**
## Description

Implements a shared `Reconcile` functionality that can be adopted by all controllers to have standard base reconciliation loop.

- Handles Adding, Removing common finalizer `greenhouse.sap/cleanup` and removes the need for individual finalizers for each resource
-  Splits the events into `EnsureCreated` and `EnsureDeleted`
- Implements standard Deletion condition to check if resource has been deleted to allow garbage collection
- Stores original resource before modification into ctx
- Does a status patch by merging the changes done to the object during reconciliation (Addition / Modification of status fields) from original object (old). This prevents resource generation from changing and can be used in Controller Mgr `WithEventFilter` to avoid duplicate reconciliation
- Implements default conditions for create state if a specific status func is not passed to `Reconcile`

## What type of PR is this? (check all applicable)

- [x] 🍕 Feature
- [ ] 🐛 Bug Fix
- [ ] 📝 Documentation Update
- [ ] 🎨 Style
- [ ] 🧑‍💻 Code Refactor
- [ ] 🔥 Performance Improvements
- [ ] ✅ Test
- [ ] 🤖 Build
- [ ] 🔁 CI
- [ ] 📦 Chore (Release)
- [ ] ⏩ Revert

## Related Tickets & Documents

- Related to #414 

## Added tests?

- [x] 👍 yes
- [ ] 🙅 no, because they aren't needed
- [ ] 🙋 no, because I need help
- [ ] Separate ticket for tests # (issue/pr)

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

## Added to documentation?

- [ ] 📜 README.md
- [ ] 🤝 Documentation pages updated
- [ ] 🙅 no documentation needed
- [ ] (if applicable) generated OpenAPI docs for CRD changes

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Review by IvoGoman — CHANGES_REQUESTED** [review]
This looks good and will make the controllers more readable/ better to maintain, also for new ones.

File: pkg/lifecycle/reconcile.go [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+		return ctrl.Result{}, err
+	}
+	//https://github.com/kubernetes/kubernetes/issues/3030
+	if runtimeObject.GetObjectKind().GroupVersionKind() == (schema.GroupVersionKind{}) {
+		err := addTypeInformationToObject(runtimeObject)
+		if err != nil {
+			logger.Error(err, "GVK is missing in runtime object")
+			return ctrl.Result{}, err
+		}
+	}
```
[the remark this answers is not in the document]
Comment (abhijith-darshan) [reply pkg/lifecycle/reconcile.go]: yes for `client-go` this is not a probelm as per the comments in the issue I will remove it

File: pkg/lifecycle/utils.go [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+	greenhousev1alpha1 "github.com/cloudoperators/greenhouse/pkg/apis/greenhouse/v1alpha1"
+)
+
+// hasCleanupFinalizer - returns true in case the cleanup finalizer exists in runtimeObject
+func hasCleanupFinalizer(runtimeObject RuntimeObject) bool {
+	return slices.Contains(runtimeObject.GetFinalizers(), CommonCleanupFinalizer)
+}
+
+// addFinalizer - Adds greenhouse cleanup finalizer to resource
+func addFinalizer(ctx context.Context, kubeClient client.Client, runtimeObject RuntimeObject) (ctrl.Result, error) {
```
[the remark this answers is not in the document]
Comment (abhijith-darshan) [reply pkg/lifecycle/utils.go:23]: the finalizer method here is doing more than an update of the finalizer on the resource I don't think we need this anymore as it is handling status patches which is not needed initially as it never hits the `EnsureCreated` or `EnsureDeleted` methods

Could we instead deprecate and remove clientutil finalizers methods gradually once everything is standardizes via `lifecycle` pkg?

File: pkg/lifecycle/utils.go [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+	if controllerutil.AddFinalizer(runtimeObject, CommonCleanupFinalizer) {
+		return ctrl.Result{}, kubeClient.Update(ctx, runtimeObject)
+	}
+	return ctrl.Result{}, nil
+}
+
+// removeFinalizer - Removes greenhouse cleanup finalizer from resource
+func removeFinalizer(ctx context.Context, kubeClient client.Client, runtimeObject RuntimeObject) (ctrl.Result, error) {
```
[the remark this answers is not in the document]
Comment (abhijith-darshan) [reply pkg/lifecycle/utils.go:31]: same here - https://github.com/cloudoperators/greenhouse/pull/623#discussion_r1792270744

File: pkg/lifecycle/utils.go [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+	deleteCondition := status.GetConditionByType(greenhousev1alpha1.DeleteCondition)
+	if deleteCondition == nil {
+		return false
+	}
+	return deleteCondition.IsTrue()
+}
+
+// patchStatus - patches the status of the resource with the new status and returns the reconcile error
+// TODO: implement event recorder to fire ready condition change events
+func patchStatus(ctx context.Context, newObject RuntimeObject, kubeClient client.Client, reconcileError error) error {
```
[the remark this answers is not in the document]
Comment (abhijith-darshan) [reply pkg/lifecycle/utils.go:52]: We are storing the old object in the context and doing a merge patch from old and new which would have necessary status conditions, plus the context methods are not exported (and it shouldn't be exported)

Could we deprecate these and remove it once everything is standardizes via `lifecycle` pkg?

File: pkg/lifecycle/reconcile.go [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+
+func ensureCreated(ctx context.Context, reconciler Reconciler, runtimeObject RuntimeObject) (ctrl.Result, error) {
+	result, reconcileResult, err := reconciler.EnsureCreated(ctx, runtimeObject)
+	convertResultToCondition(runtimeObject, reconcileResult, true)
+	return result, err
+}
+
+func ensureDeleted(ctx context.Context, reconciler Reconciler, runtimeObject RuntimeObject) (ctrl.Result, error) {
+	result, reconcileResult, err := reconciler.EnsureDeleted(ctx, runtimeObject)
+	convertResultToCondition(runtimeObject, reconcileResult, false)
```
[the remark this answers is not in the document]
Comment (abhijith-darshan) [reply pkg/lifecycle/reconcile.go]: sounds good, I will do the necessary changes

[the remark this answers is not in the document]
Comment (abhijith-darshan) [reply pkg/lifecycle/reconcile.go]: If deletion fails due to some reason then we would not be able to report that... The resource conditions may continue to show old conditions... Ready 'True'....

I thought about resetting all Conditions when deletion is invoked and show only deletion condition and report errors in there that way if deletion fails the error is reported there.

[the remark this answers is not in the document]
Comment (abhijith-darshan) [reply pkg/lifecycle/reconcile.go]: as discussed re-purposed `DeletionScheduled ConditionType` to `Delete ConditionType` with Reason to describe whether it is a scheduled deletion or an ongoing deletion

File: pkg/clientutil/finalizer_test.go [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+	}
+
+	for _, test := range testCases {
+		t.Run(test.name, func(t *testing.T) {
+			if got := HasFinalizer(test.args.runtimeObject, test.args.finalizer); got != test.want {
+				t.Errorf("HasFinalizer() = %v, want %v", got, test.want)
+			}
+		})
+	}
+}
```
[the remark this answers is not in the document]
Comment (abhijith-darshan) [reply pkg/clientutil/finalizer_test.go:78]: Ok will remove it

File: pkg/clientutil/finalizer.go [hunk 7]
```diff
@@ -33,3 +33,7 @@ func RemoveFinalizer(ctx context.Context, c client.Client, o client.Object, fina
 	})
 	return err
 }
+
+func HasFinalizer(o client.Object, finalizer string) bool {
+	return controllerutil.ContainsFinalizer(o, finalizer)
+}
```
[the remark this answers is not in the document]
Comment (abhijith-darshan) [reply pkg/clientutil/finalizer.go:39]: Will remove it.

File: pkg/lifecycle/utils.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (abhijith-darshan) [reply pkg/lifecycle/utils.go]: If this is ok then would need you to re-review the new changes and discard the change request for this

File: pkg/lifecycle/reconcile.go [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+		}
+		condition = greenhousev1alpha1.FalseCondition(greenhousev1alpha1.ReadyCondition, FailingCreationReason, "resource creation failed"+msg)
+	default:
+		condition = greenhousev1alpha1.UnknownCondition(greenhousev1alpha1.ReadyCondition, PendingCreationReason, "resource creation is pending")
+	}
+	runtimeObject.SetCondition(condition)
+}
+
+// patchStatus - patches the status of the resource with the new status and returns the reconcile error
+// TODO: implement event recorder to fire ready condition change events
```
Comment (ibakshay) [inline pkg/lifecycle/reconcile.go:174]: Great work Abhi! 🚀  I went thru your PR to keep myself updated
Have only one question.
Is this TODO still relevant for this PR?

Comment (abhijith-darshan) [reply pkg/lifecycle/reconcile.go:174]: Hi @ibakshay No not for this PR, there will be a follow up task, can you give me a  `suggestion` to remove the comment?

Comment (ibakshay) [inline pkg/lifecycle/reconcile.go:174]:
```suggestion
```

**Review by IvoGoman — APPROVED** [review]

**Pull request merged**